src/recon_buildblock/ML_estimate_component_based_normalisation.cxx

```diff
@@ -84,225 +127,274 @@ ML_estimate_component_based_normalisation(const std::string& out_filename_prefix
-
-  GeoData3D measured_geo_data(num_physical_axial_crystals_per_basic_unit,
-                              num_physical_transaxial_crystals_per_basic_unit / 2,
-                              num_physical_rings,
-                              num_physical_detectors_per_ring); // inputes have to be modified
-  GeoData3D norm_geo_data(num_physical_axial_crystals_per_basic_unit,
-                          num_physical_transaxial_crystals_per_basic_unit / 2,
-                          num_physical_rings,
-                          num_physical_detectors_per_ring); // inputes have to be modified
-
-  BlockData3D measured_block_data(num_axial_blocks, num_transaxial_blocks, num_axial_blocks - 1, num_transaxial_blocks - 1);
-  BlockData3D norm_block_data(num_axial_blocks, num_transaxial_blocks, num_axial_blocks - 1, num_transaxial_blocks - 1);
-
-  make_fan_data_remove_gaps(model_fan_data, model_data);
-  {
-    // next could be local if KL is not computed below
-    FanProjData measured_fan_data;
-    float threshold_for_KL;
-    // compute factors dependent on the data
+  // Setup the data structures given the PET scanner geometry
+  data_fan_sums = DetectorEfficiencies(IndexRange2D(num_physical_rings, num_physical_detectors_per_ring));
+  efficiencies_ptr = std::make_shared<DetectorEfficiencies>(IndexRange2D(num_physical_rings, num_physical_detectors_per_ring));
+
+  measured_geo_data = GeoData3D(num_physical_axial_crystals_per_basic_unit,
+                                num_physical_transaxial_crystals_per_basic_unit / 2,
+                                num_physical_rings,
+                                num_physical_detectors_per_ring); // inputes have to be modified
+  norm_geo_data_ptr = std::make_shared<GeoData3D>(num_physical_axial_crystals_per_basic_unit,
+                                                  num_physical_transaxial_crystals_per_basic_unit / 2,
+                                                  num_physical_rings,
+                                                  num_physical_detectors_per_ring); // inputes have to be modified
+
+  measured_block_data = BlockData3D(num_axial_blocks, num_transaxial_blocks, num_axial_blocks - 1, num_transaxial_blocks - 1);
+  norm_block_data_ptr
+      = std::make_shared<BlockData3D>(num_axial_blocks, num_transaxial_blocks, num_axial_blocks - 1, num_transaxial_blocks - 1);
+
+  make_fan_data_remove_gaps(model_fan_data, *model_projdata_sptr);
+  make_fan_data_remove_gaps(measured_fan_data, *measured_projdata_sptr);
+
+  threshold_for_KL = compute_threshold_for_KL();
+
+  make_fan_sum_data(data_fan_sums, measured_fan_data);
+  make_geo_data(measured_geo_data, measured_fan_data);
+  make_block_data(measured_block_data, measured_fan_data);
+  if (do_display)
     {
-      make_fan_data_remove_gaps(measured_fan_data, measured_data);
+      display(measured_block_data, "raw block data from measurements");
+    }
 
-      /* TEMP FIX */
```

Comment:
I can't remember where this came from and would have to review, but are you sure we can cut it?

Reply:
This `TEMP FIX` does 
```
if (model_fan_data(ra, a, rb, b) == 0)
    measured_fan_data(ra, a, rb, b) = 0;
```
and then sets the `threshold_for_KL = measured_fan_data.find_max() / 100000.F;`

This isn't removed, refactored into `compute_threshold_for_KL`. GitHub is not displaying these changes very well.

Reply:
I haven't checked, but this fix will not just be important for KL, but probably also for the iterative update. (it'll detect 0/0, as oposed 1/0).